shared/actions/profile.js

```diff
@@ -42,3 +49,91 @@ export function editProfile (bio: string, fullname: string, location: string) :
     })
   }
 }
+
+function makeWaitingHandler (dispatch: Dispatch): {waitingHandler: (waiting: boolean) => void} {
+  return {
+    waitingHandler: bindActionCreators(waitingForResponse, dispatch),
+  }
+}
+
+function waitingForResponse (waiting: boolean): Waiting {
+  return {
+    type: Constants.waiting,
+    payload: {waiting},
+  }
+}
+
+function selectPlatform (platform: PlatformsExpanded): UpdatePlatform {
+  return {
+    type: Constants.updatePlatform,
+    payload: {platform},
+  }
+}
+
+let submitUsernameResponse: ?Object = null
+
+function submitUsername (): AsyncAction {
+  return (dispatch, getState) => {
+    if (submitUsernameResponse) {
+      submitUsernameResponse.result(getState().profile.username)
+      submitUsernameResponse = null
+    }
+  }
+}
+
+function updateUsername (username: string): UpdateUsername {
+  return {
+    type: Constants.updateUsername,
+    payload: {username},
+  }
+}
+
+function cancelAddProof () : AsyncAction {
+  return (dispatch) => {
+    if (submitUsernameResponse) {
+      engine.cancelRPC(submitUsernameResponse, InputCancelError)
+      submitUsernameResponse = null
+    }
+    dispatch(navigateUp())
+  }
+}
+
+function addProof (platform: PlatformsExpanded): AsyncAction {
+  return (dispatch) => {
+    dispatch(selectPlatform(platform))
+
+    proveStartProofRpc({
+      ...makeWaitingHandler(dispatch),
+      param: {
+        service: platform,
+        username: '',
+        force: true,
+        promptPosted: false,
+        auto: false,
+      },
+      incomingCallMap: {
+        'keybase.1.proveUi.promptUsername': ({prompt, prevError}, response) => {
+          submitUsernameResponse = response
+          dispatch(routeAppend({path: 'ProveEnterUsername'}))
```

Comment:
using the string based navigation stuff / entirely store based action except for the stashing of the response object